### PR TITLE
feat: allow operation handlers to match on json payloads

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes
The extension of custom operation handler makes it easy to use c8y custom operations in thin-edge. Users no longer have to create custom SmartREST templates.

Changes:
- custom operation handler can accept JSON over MQTT topic
- command field accepts arguments from JSON payload, same format as operation workflow
- supporting `504`-`506` messages to update operation status
- added `skip_status_update = true` option

Todo:
- [x] add `skip_status_update` option
- [x] error handling
- [x] unit/system test

Will be done in follow-up:
- [ ] user guide
- [ ] quick refactoring https://github.com/thin-edge/thin-edge.io/pull/3144#discussion_r1785941656

---

**Example usage:**

Install [`c8y-command-plugin`](https://github.com/thin-edge/c8y-command-plugin) and modify `/etc/tedge/operations/c8y/c8y_Command` as below.

```toml
[exec]
topic = "c8y/devicecontrol/notifications"
on_fragment = "c8y_Command"
command = "/usr/bin/c8y-command ${.payload.c8y_Command.text}"
skip_status_update = true
```

Then, try **Shell** tab in c8y.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#3095 "Drop 1"

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_command/command_handler.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    echo "$(date --iso-8601=seconds 2>/dev/null || date -Iseconds) : INFO : $*" >&2
+}
+
+# Parse the message
+COMMAND="${1}"
+
+# Check if command is wrapped with quotes, if so then remove them
+if [[ "$COMMAND" == \"*\" ]]; then
+    COMMAND="${COMMAND:1:-1}"
+fi
```
Comment (didier-wenzek) [inline tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_command/command_handler.sh:13]: Since you are running the command using `bash -c "$COMMAND"`, this check is useless.

```suggestion
```

File: crates/extensions/c8y_mapper_ext/src/tests.rs [hunk 2]
```diff
@@ -1947,6 +1948,57 @@ EOF
     assert_command_exec_log_content(ttd, expected_content);
 }
 
+#[ignore] // HELP!
```
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/tests.rs:1951]: ;-) A concrete question might be helpful to get some help.

Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/tests.rs:1951]: More seriously: https://github.com/thin-edge/thin-edge.io/pull/3144#discussion_r1786102656

File: crates/core/c8y_api/src/smartrest/operations.rs [hunk 3]
```diff
@@ -25,14 +23,16 @@ pub enum ResultFormat {
 /// Operations are derived by reading files subdirectories per cloud /etc/tedge/operations directory
 /// Each operation is a file name in one of the subdirectories
 /// The file name is the operation name
-
 #[derive(Debug, Clone, Deserialize, Eq, PartialEq)]
 #[serde(rename_all = "snake_case")]
 pub struct OnMessageExec {
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/c8y_api/src/smartrest/operations.rs:28]: I agree, but I prefer to do as you suggested after this PR, otherwise moving struct and changes will be mixed.

File: crates/core/c8y_api/src/smartrest/operations.rs [hunk 4]
```diff
@@ -165,6 +177,35 @@ impl Operations {
...
+                        warn!(
+                            "'on_message' should not be provided in the custom operation handler mapping '{}', which is for SmartREST input.",
+                            op.name
+                        );
+                    } else if op.on_fragment().is_none() {
+                        warn!(
+                            "'on_fragment' is missing in the custom operation handler mapping '{}'",
+                            op.name
+                        );
+                    } else {
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/c8y_api/src/smartrest/operations.rs:200]: I was actually also considering the way. I'll try to implement a validation function quickly after I finish doc. It should cover other patterns as well.

For all:
- files without `[exec]` field
For SmartREST handlers:
- `topic` is missing
- `on_message` is missing
- `on_fragment` is given
- `command` is missing
For JSON handlers:
- `on_message` is given
- `on_fragment` is missing
- `command` is missing (this will be changed once "drop 2" is completed

The good thing of a centralized validation function is, when we need to add something new, developer just has to revisit the function. If validation is made in distributed functions, one needs to revisit all of them.

File: crates/extensions/c8y_mapper_ext/src/tests.rs [hunk 5]
```diff
@@ -1947,6 +1948,57 @@ EOF
...
+            "externalSource":{
+           "externalId":"TST_haul_searing_set",
+           "type":"c8y_Serial"
+        }
+             })
+        .to_string(),
+    );
+    mqtt.send(input_message).await.expect("Send failed");
+
+    let aaa = mqtt.recv().await; // FIXME: why None??
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/tests.rs:1992]: Actually the 1) was the root cause! 2) is not an issue. 3) needs tedge config `smartrest_use_operation_id` to be `true`. Now fixed. Thank you!

File: crates/extensions/c8y_mapper_ext/src/tests.rs (hunk 2 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/tests.rs]: fixed in e38f515cac74323ade9e64c7af1099db6cc5642e

File: crates/core/c8y_api/src/smartrest/operations.rs [hunk 6]
```diff
@@ -165,6 +177,35 @@ impl Operations {
         None
     }
 
+    pub fn filter_for_json_input(&self, topic_name: &str) -> Vec<Operation> {
+        let mut vec: Vec<Operation> = Vec::new();
+        for op in self.operations.clone() {
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/c8y_api/src/smartrest/operations.rs]: fixed in e38f515cac74323ade9e64c7af1099db6cc5642e

File: crates/extensions/c8y_mapper_ext/src/tests.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/tests.rs]: fixed in e38f515cac74323ade9e64c7af1099db6cc5642e

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 7]
```diff
@@ -706,15 +717,64 @@ impl CumulocityConverter {
...
+
+                if handlers.is_empty() {
+                    info!("No matched custom operation handler is found for the topic {json_over_mqtt_topic}. The operation '{operation_id}' (ID) is ignored.");
+                }
+
+                for custom_handler in &handlers {
+                    let on_fragment = custom_handler.on_fragment().unwrap(); // already checked while filtering
+                    if extras.contains_key(&on_fragment) {
+                        self.execute_custom_operation(custom_handler, message, &operation_id)
+                            .await?;
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:731]: Given that `on_fragment = "c8y_Commnand"`, your JSON input is not an error. Even in c8y level, you are allowed to give many custom fragments in an operation payload. 
The custom operation handler reacts when `c8y_Command` exists in the JSON input. So `on_fragment` works like "match".

User can even define like below. Freedom of choice what he wants to do.
```toml
command = "/usr/bin/something.sh ${.payload.c8y_Another_Command.text}"
```

File: crates/core/c8y_api/src/smartrest/operations.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/c8y_api/src/smartrest/operations.rs]: Done in f9ef26492cd7f7b264ce652b4998c626f4b1926c

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 8]
```diff
@@ -706,15 +717,64 @@ impl CumulocityConverter {
...
-                // However, these operations can be addressed by SmartREST that is published together with JSON over MQTT
+                let json_over_mqtt_topic = C8yDeviceControlTopic::name(&self.config.c8y_prefix);
+                let handlers = self.operations.filter_for_json_input(&json_over_mqtt_topic);
+
+                if handlers.is_empty() {
+                    info!("No matched custom operation handler is found for the topic {json_over_mqtt_topic}. The operation '{operation_id}' (ID) is ignored.");
+                }
+
+                for custom_handler in &handlers {
+                    let on_fragment = custom_handler.on_fragment().unwrap(); // already checked while filtering
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs]: Done in f9ef26492cd7f7b264ce652b4998c626f4b1926c

File: crates/core/c8y_api/src/smartrest/operations.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/core/c8y_api/src/smartrest/operations.rs:28]: This would help. The file is currently organized upside down, from the details to the main point.

File: crates/core/c8y_api/src/smartrest/operations.rs [hunk 9]
```diff
@@ -119,6 +147,55 @@ impl Operation {
             .map(|exec| exec.forceful_timeout)
             .unwrap_or(DEFAULT_FORCEFUL_TIMEOUT)
     }
+
+    pub fn is_valid_opeation_handler(&self) -> bool {
+        if self.exec.is_none() {
+            return true; // can be an empty operation file, not meant for custom operation handler
```
Comment (didier-wenzek) [inline crates/core/c8y_api/src/smartrest/operations.rs:153]: Returning false makes more sense: this is not an operation handler.

File: crates/core/c8y_api/src/smartrest/operations.rs [hunk 10]
```diff
@@ -119,6 +147,55 @@ impl Operation {
             .map(|exec| exec.forceful_timeout)
             .unwrap_or(DEFAULT_FORCEFUL_TIMEOUT)
     }
+
+    pub fn is_valid_opeation_handler(&self) -> bool {
```
Comment (didier-wenzek) [inline crates/core/c8y_api/src/smartrest/operations.rs:151]:
```suggestion
    pub fn is_valid_operation_handler(&self) -> bool {
```

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:730]: > Given that `on_fragment = "c8y_Commnand"`, your JSON input is not an error. Even in c8y level, you are allowed to give many custom fragments in an operation payload. 
The custom operation handler reacts when `c8y_Command` exists in the JSON input. So `on_fragment` works like "match".

I don't fully understand your answer.

What I get, is that:
- A c8y command payload can have several `c8y_*` fragments.
- A thin-edge operation handler can be defined for each of these `c8y_*` fragments.
- If there is a matching thin-edge operation handler for all the c8y_*` fragments of the c8y command,
  then, *all* these commands will be executed in sequence.
- *But*, c8y might be confused to receive several command status for the same command id.

In other words, I'm surprised that the code triggers *all* the matching command handlers.

File: crates/core/c8y_api/src/smartrest/operations.rs (hunk 9 above)
Comment (rina23q) [reply crates/core/c8y_api/src/smartrest/operations.rs:153]: Yes, but `CumulocityConverter::operations` needs to contain **both** custom operation handlers and empty operation declaration files. 
https://github.com/thin-edge/thin-edge.io/blob/main/crates/extensions/c8y_mapper_ext/src/converter.rs#L1425

I have an idea. I'll create another function, `is_supported_operation_file()`, then I'll change this like https://github.com/thin-edge/thin-edge.io/blob/f9ef26492cd7f7b264ce652b4998c626f4b1926c/crates/core/c8y_api/src/smartrest/operations.rs#L306-L308 to

```rust
            if details.is_valid_opeation_handler() || details.is_supported_operation_file() {
                operations.add_operation(details);
            }
```

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:730]: > But, c8y might be confused to receive several command status for the same command id.

This is a very valid question, that I didn't think of. @reubenmiller Can you help us here what is the expected behaviour for the Didier's scenario? The first matched command handler only should be executed?

[the remark this answers is not in the document]
Comment (reubenmiller) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:730]: Generally, Cumulocity IoT will only have one unique identifying fragment (like `c8y_Command`) which is related to a specific command. So if someone has done the Cumulocity IoT data design correctly then it *shouldn't* be an issue.

However we can improve on the matching mechanism in the future, and also allow a key/value style matching in addition to the "key exists" logic that we currently use.

File: tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_command/command_handler.sh (hunk 1 above)
Comment (rina23q) [reply tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_command/command_handler.sh]: Done in 934f052

File: crates/core/c8y_api/src/smartrest/operations.rs (hunk 9 above)
Comment (rina23q) [reply crates/core/c8y_api/src/smartrest/operations.rs]: Done in 31913a47803f14de3bb1d2c6c5089cf223f3df14

File: crates/core/c8y_api/src/smartrest/operations.rs (hunk 10 above)
Comment (rina23q) [reply crates/core/c8y_api/src/smartrest/operations.rs]: Oops. Done in 31913a47803f14de3bb1d2c6c5089cf223f3df14

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:730]: @reubenmiller 
I know it's not a desired pattern, however, If there two files that has the same value on `on_fragment`, both handlers should be executed (current implementation)? or it should limit one?

file: `/etc/tedge/operations/c8y/c8y_Custom1`
```
[exec]
topic = "c8y/devicecontrol/notifications"
on_fragment = "c8y_Custom"
command = "/usr/bin/something.sh ${c8y_Custom.something}"
```

file: `/etc/tedge/operations/c8y/c8y_Custom2`
```
[exec]
topic = "c8y/devicecontrol/notifications"
on_fragment = "c8y_Custom"
command = "/usr/bin/something.sh ${c8y_Custom.something}"
```

[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:730]: Yep, done in 024813d286a8ca5767d5eb26ccbf6ddfd0e4a07d

**Review by didier-wenzek — APPROVED** [review]
Approved

**Comment by rina23q** [comment]
If there is a custom operation handlers with invalid contents, you may see multiple times of the same warning message as below.

```sh
WARN c8y_api::smartrest::operations: ''command' is missing in the JSON custom operation handler mapping 'my_CustomOperation'
```

This is caused by too many loadings of the `/etc/tedge/operations/c8y` directory by c8y-mapper. Reported in https://github.com/thin-edge/thin-edge.io/issues/3160.

**Comment by gligorisaev** [comment]
Review done, implemented as described, the tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_command/custom_operation_command.robot
is doing what it is intended for. Checked for flakines.
